src/backend/opencl/jit.cpp

```diff
@@ -48,108 +47,128 @@ static string getFuncName(Node *node, bool is_linear, bool *is_double)
         funcName << "G_";
     }
 
-    std::string outName = node->getNameStr();
-    funcName << outName;
-
-    node->genKerName(funcName);
-
-    string nameStr = funcName.str();
-    funcName << nameStr;
+    bool is_dbl = false;
+    int id = 0;
+    for (auto node :  nodes) {
+        id = node->setId(id);
+        std::string outName = node->getNameStr();
+        funcName << outName;
+
+        node->genKerName(funcName);
+        string nameStr = funcName.str();
+        funcName << nameStr;
+
+        nameStr = nameStr + outName;
+        string dblChars = "dDzZ";
+        size_t loc = nameStr.find_first_of(dblChars);
+        is_dbl |= (loc != std::string::npos);
+    }
 
-    nameStr = nameStr + outName;
-    string dblChars = "dDzZ";
-    size_t loc = nameStr.find_first_of(dblChars);
-    *is_double = (loc != std::string::npos);
+    *is_double = is_dbl;
 
     std::hash<std::string> hash_fn;
     hashName << "KER" << hash_fn(funcName.str());
     return hashName.str();
 }
 
-static string getKernelString(string funcName, Node *node, bool is_linear)
+static string getKernelString(string funcName, std::vector<Node *> nodes, bool is_linear)
 {
-    stringstream kerStream;
-    int id = node->getId();
 
-    kerStream << "__kernel void" << "\n";
+    // Common OpenCL code
+    // This part of the code does not change with the kernel.
+
+    static const char *kernelVoid =  "__kernel void\n";
+    static const char *dimParams = "KParam oInfo, uint groups_0, uint groups_1, uint num_odims";
+    static const char *blockStart = "{\n\n";
+    static const char *blockEnd = "\n\n}";
+
+    static const char *linearIndex = "\n"
+        "uint groupId  = get_group_id(1) * get_num_groups(0) + get_group_id(0);\n"
+        "uint threadId = get_local_id(0);\n"
+        "int idx = groupId * get_local_size(0) * get_local_size(1) + threadId;\n"
+        "if (idx >= oInfo.dims[3] * oInfo.strides[3]) return;\n";
+
+    static const char *generalIndex = "\n"
```

Comment:
Yeah, I wanted to do that as well. But my emacs config did not have C++11 support :p I did not want to screw up the indentation.